core/src/main/scala/kafka/coordinator/group/PartitionWriterImpl.scala

```diff
@@ -0,0 +1,144 @@
+import org.apache.kafka.common.record.{CompressionType, MemoryRecords, TimestampType}
+import org.apache.kafka.common.record.Record.EMPTY_HEADERS
+import org.apache.kafka.common.utils.Time
+import org.apache.kafka.coordinator.group.runtime.PartitionWriter
+import org.apache.kafka.storage.internals.log.AppendOrigin
+
+import java.nio.ByteBuffer
+import java.util
+import scala.collection.Map
+
+private[group] class ListenerAdaptor(
+  val listener: PartitionWriter.Listener
+) extends PartitionListener {
+  override def onHighWatermarkUpdated(
+    tp: TopicPartition,
+    offset: Long
+  ): Unit = {
+    listener.onHighWatermarkUpdated(tp, offset)
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case other: ListenerAdaptor => listener.equals(other.listener)
+    case _ => false
+  }
+
+  override def hashCode(): Int = {
+    listener.hashCode()
+  }
+
+  override def toString: String = {
+    s"ListenerAdaptor(listener=$listener)"
+  }
+}
+
+class PartitionWriterImpl[T](
+  replicaManager: ReplicaManager,
+  serializer: PartitionWriter.Serializer[T],
+  compressionType: CompressionType,
+  time: Time
+) extends PartitionWriter[T] {
+
+  override def registerListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.maybeAddListener(tp, new ListenerAdaptor(listener))
+  }
+
+  override def deregisterListener(
+    tp: TopicPartition,
+    listener: PartitionWriter.Listener
+  ): Unit = {
+    replicaManager.removeListener(tp, new ListenerAdaptor(listener))
+  }
+
+  override def append(
+    tp: TopicPartition,
+    records: util.List[T]
+  ): Long = {
+    if (records.isEmpty) {
+      throw new IllegalStateException("records must be non-empty.")
+    }
+
+    replicaManager.getLogConfig(tp) match {
+      case Some(logConfig) =>
+        val magic = logConfig.recordVersion.value
+        val maxBatchSize = logConfig.maxMessageSize
+        val currentTimeMs = time.milliseconds()
+
+        val recordsBuilder = MemoryRecords.builder(
+          ByteBuffer.allocate(math.min(16384, maxBatchSize)),
+          magic,
+          compressionType,
+          TimestampType.CREATE_TIME,
+          0L,
+          maxBatchSize
+        )
+
+        records.forEach { record =>
+          val keyBytes = serializer.serializeKey(record)
+          val valueBytes = serializer.serializeValue(record)
+
+          if (recordsBuilder.hasRoomFor(currentTimeMs, keyBytes, valueBytes, EMPTY_HEADERS)) {
+            recordsBuilder.append(
+              currentTimeMs,
+              keyBytes,
+              valueBytes,
+              EMPTY_HEADERS
+            )
+          } else {
+            throw new RecordTooLargeException(s"Message batch size is ${recordsBuilder.estimatedSizeInBytes()} bytes " +
+              s"in append to partition $tp which exceeds the maximum configured size of $maxBatchSize.")
+          }
+        }
+
+        val appendResults = replicaManager.appendToLocalLog(
```

Comment:
Are the partitions we are listening to just the __consumer_offsets partitions? 
I think this class is pretty generic (which is fine), but this append origin being the coordinator makes it less generic.

I know we are in the coordinator/group path, but it might not be obvious without looking at the file path.

Reply:
Yeah, I agree. The alternative would be to pass it via the method but this would mean hardcoding it somewhere else. Is it better?

Reply:
I'm not sure if it is maybe better to include a comment or include it somehow in the name. Hard coding is fine as long as it isn't a surprise 😅

Reply:
Let me rename the class to `CoordinatorPartitionWriter` to make it explicit that this is used for/by coordinators. Does this work for you?

Reply:
Sounds good. Thanks!